### PR TITLE
Add description and title to consumables

### LINKED ISSUE #1276: AFR-639 Addition to `!cc create`: `-title` and `-desc`
**Feature Request:** Addition to `!cc create`: `-title` and `-desc`
**Who would use it?** Everyone
**What part of the bot?** Custom counters (so character model?)
**How would it work?** Specifying `-title "{name} does a Thing!"` and `-desc "Mechanics Text"` when running `!cc create`, and then having the title and desc display when using `!cc <counter> -1`
**Why should this be added?** Currently, you need to write an alias to display any text with cc manipulation. Having this just be built in to `!cc` will make it easier for users to set up basic non-automated things.

For example, Divine Sense and War Priest don't have any automation, and just marking off a use and displaying mechanical text would handle them quite well.

- silverbass#2407

> Upvote - Dr Turtle#1771

> Upvote - spamandtuna#9657

> Upvote - Lathaon#6649

> Upvote - Spooky Riptidecord#7191

> Upvote - Hopop#6751

> Upvote - Derixyleth#0636

> Upvote - Zekl#1505

> Upvote - Robbe#7964

> Upvote - Roedelius#9141

> Upvote - Mobius#1442

> Upvote - Ithala#3408

> Upvote - syrup#1965

Votes: +12 / -0

### PR DESCRIPTION
**Pull request by 1drturtle**
Summary
======
Adds a description and title to custom counters/consumables
In the title, `[name]` is replaced with the character name and `[cname]` is replace with the custom counter name.

Resolves #1276 (AFR-639)

Methods added
----
`AliasCustomCounter` -
* `AliasCustomCounter.title`
* `AliasCustomCounter.desc`

`CustomCounter` -
* `CustomCounter.title`
* `CustomCounter.desc`

Checklist
======

#### PR Type

- [X] This PR is a code change that implements a feature request.
- [ ] This PR fixes an issue.
- [ ] This PR adds a new feature that is not an open feature request.
- [ ] This PR is not a code change (e.g. documentation, README, ...)
#### Other
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [X] If code changes were made then they have been tested.
- [X] I have updated the documentation to reflect the changes.

### CODE REVIEW

File: cogs5e/models/sheet/player.py [hunk 1]
```diff
@@ -94,9 +94,13 @@ class CustomCounter:
                  'hp': "Gaining HP"}
 
     def __init__(self, character, name, value, minv=None, maxv=None, reset=None, display_type=None, live_id=None,
-                 reset_to=None, reset_by=None):
+                 title=None, desc=None, reset_to=None, reset_by=None):
```
Comment (mommothazaz123) [inline cogs5e/models/sheet/player.py:97]:
```suggestion
                 reset_to=None, reset_by=None, title=None, desc=None):
```

I don't think `__init__` is ever called with only positional args, but this prevents breaking it if it is.

File: cogs5e/models/sheet/player.py [hunk 2]
```diff
@@ -116,12 +120,12 @@ def from_dict(cls, char, d):
...
         return {"name": self.name, "value": self._value, "minv": self.min, "maxv": self.max, "reset": self.reset_on,
-                "display_type": self.display_type, "live_id": self.live_id, "reset_to": self.reset_to,
-                "reset_by": self.reset_by}
+                "display_type": self.display_type, "live_id": self.live_id, "title": self.title, "desc": self.desc,
+                "reset_to": self.reset_to, "reset_by": self.reset_by}
 
     @classmethod
     def new(cls, character, name, minv=None, maxv=None, reset=None, display_type=None, live_id=None,
-            reset_to=None, reset_by=None):
+            title=None, desc=None, reset_to=None, reset_by=None):
```
Comment (mommothazaz123) [inline cogs5e/models/sheet/player.py:128]:
```suggestion
            reset_to=None, reset_by=None, title=None, desc=None):
```

File: cogs5e/models/sheet/player.py [hunk 3]
```diff
@@ -168,7 +172,18 @@ def new(cls, character, name, minv=None, maxv=None, reset=None, display_type=Non
...
+            raise InvalidArgument('Description must be less than 800 characters.')
+
+        if title and len(title) >= 256:
+            raise InvalidArgument('Title must be less than 256 characters.')
+
+        if len(name) > 256:
+            raise InvalidArgument('Name must be less than 256 characters.')
+
+        return cls(character, name.strip(), initial_value, minv, maxv, reset, display_type, live_id,
+                   title, desc, reset_to, reset_by)
```
Comment (mommothazaz123) [inline cogs5e/models/sheet/player.py:186]:
```suggestion
                   reset_to, reset_by, title, desc)
```
to match the positional arg suggestions

File: cogs5e/models/sheet/player.py [hunk 4]
```diff
@@ -196,6 +211,25 @@ def get_reset_to(self):
     def value(self):
         return self._value
 
+    @property
```
Comment (mommothazaz123) [inline cogs5e/models/sheet/player.py:214]: I'm not entirely sure these properties (and similarly, `_title` and `_desc` being private attrs) are entirely necessary.

As is, `CustomCounter.title` (the property here) is what is called when it's committed to the database, so the `[cname]` and `[name]`s are lost - I don't believe `[cname]` and `[name]` are necessary, though.

File: cogs5e/models/sheet/player.py [hunk 5]
```diff
@@ -274,14 +312,15 @@ def full_str(self):
     def __str__(self):
         _max = self.get_max()
 
+        out = self.desc + '\n' if self.desc else ''
```
Comment (mommothazaz123) [inline cogs5e/models/sheet/player.py:315]: This'll put the description (up to 800 characters) in *every* location the counter is stringified. That doesn't seem necessary - for example, if I had an alias 
```
!alias wildmagic embed <drac2>
out = [...]
# some wild magic code here
out.append(f'-f "Some Resource|{{character().cc_str('Some Resource')}}"')
return '\n'.join(out)
```

I wouldn't want the description to appear.

**Review by mommothazaz123 — CHANGES_REQUESTED** [review]
Thanks for the contribution! I'm not entirely sure what the scope of the feature request this PR is closing is. Should titles and descriptions appear everywhere? In rests? Only when they're manually changed with `!cc`?

I believe the original scope of the FR only encompasses the latter (and thus that the title and desc handling should be in that command rather than in methods on the cc object), and that the `[name]` and `[cname]` modification is unnecessary.

**Comment by 1drturtle** [comment]
Updated - 
Removed description from `__str__`
Fixed `title` and `desc` properties (`[name]` and `[cname]` are still supported, but are replaced in the command instead.)
Moved order of `title` and `desc`
image
(In aliasing)
image

RFC - Remove `[cname]` since when it's created they know the name of it?

**Comment by mommothazaz123** [comment]
Screenshot 2020-10-26 at 11 43 51

Maybe the description should have its own field (under the meta of cc info/cc value) - this looks a little cluttered

**Comment by 1drturtle** [comment]
With or without description header? I'm leaning more towards with header so I that's what I committed.
image

File: cogs5e/models/sheet/player.py [hunk 6]
```diff
@@ -252,11 +267,13 @@ def full_str(self):
...
 
+        val = ''
+
         if self.display_type == 'bubble':
             assert self.max is not None
-            val = f"{bubble_format(self.value, _max)}\n"
+            val += f"{bubble_format(self.value, _max)}\n"
         else:
-            val = f"**Current Value**: {self.value}\n"
+            val += f"**Current Value**: {self.value}\n"
```
Comment (mommothazaz123) [inline cogs5e/models/sheet/player.py:276]: This change can probably be reverted

File: cogs5e/gametrack.py [hunk 7]
```diff
@@ -434,15 +434,25 @@ async def spellbook_remove(self, ctx, *, spell_name):
...
             return await self.customcounter_summary(ctx)
         character: Character = await Character.from_ctx(ctx)
         counter = await character.select_consumable(ctx, name)
 
+        cc_embed_title = counter.title if counter.title is not None else counter.name
+
+        # replace [name] in title
+        cc_embed_title = cc_embed_title.replace('[name]', character.name)
+
+        cc_desc = counter.desc if counter.desc else ''
```
Comment (mommothazaz123) [inline cogs5e/gametrack.py:449]: The use of `cc_desc` below can probably be changed to just checks on `counter.desc`, and this variable removed

**Review by mommothazaz123 — APPROVED** [review]

**Pull request merged**